examples/fabric-admin/device_manager/PairingManager.cpp

```diff
@@ -168,3 +247,444 @@ void PairingManager::OnOpenCommissioningWindowVerifierResponse(void * context, N
+        }
+        // These Optionals must have values now.
+        // The commissioner will verify these values.
+        params.SetICDSymmetricKey(mICDSymmetricKey.Value());
+        if (mICDStayActiveDurationMsec.HasValue())
+        {
+            params.SetICDStayActiveDurationMsec(mICDStayActiveDurationMsec.Value());
+        }
+        params.SetICDCheckInNodeId(mICDCheckInNodeId.Value());
+        params.SetICDMonitoredSubject(mICDMonitoredSubject.Value());
+        params.SetICDClientType(mICDClientType.Value());
+    }
+
+    return params;
+}
+
+CHIP_ERROR PairingManager::Pair(NodeId remoteId, Transport::PeerAddress address)
+{
+    auto params = RendezvousParameters().SetSetupPINCode(mSetupPINCode).SetDiscriminator(mDiscriminator).SetPeerAddress(address);
+
+    CHIP_ERROR err           = CHIP_NO_ERROR;
+    auto commissioningParams = GetCommissioningParameters();
+    err                      = CurrentCommissioner().PairDevice(remoteId, params, commissioningParams);
+
+    return err;
+}
+
+void PairingManager::OnCurrentFabricRemove(void * context, NodeId nodeId, CHIP_ERROR err)
+{
+    PairingManager * self = reinterpret_cast<PairingManager *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogError(NotSpecified, "OnCurrentFabricRemove: context is null"));
+
+    if (err == CHIP_NO_ERROR)
+    {
+        // print to console
+        fprintf(stderr, "Device with Node ID: " ChipLogFormatX64 "has been successfully removed.\n", ChipLogValueX64(nodeId));
+
+#if defined(PW_RPC_ENABLED)
+        FabricIndex fabricIndex = self->CurrentCommissioner().GetFabricIndex();
+        app::InteractionModelEngine::GetInstance()->ShutdownSubscriptions(fabricIndex, nodeId);
+        ScopedNodeId scopedNodeId(nodeId, fabricIndex);
+        RemoveSynchronizedDevice(scopedNodeId);
+#endif
+    }
+    else
+    {
+        ChipLogProgress(NotSpecified, "Device unpair Failure: " ChipLogFormatX64 " %s", ChipLogValueX64(nodeId), ErrorStr(err));
+    }
+}
+
+void PairingManager::InitPairingCommand()
+{
+    mCommissioner->RegisterPairingDelegate(this);
+    // Clear the CATs in OperationalCredentialsIssuer
+    mCredIssuerCmds->SetCredentialIssuerCATValues(kUndefinedCATs);
+    mDeviceIsICD = false;
+}
+
+CHIP_ERROR PairingManager::PairDeviceWithCode(NodeId nodeId, const char * payload)
+{
+    if (payload == nullptr || strlen(payload) > kMaxManualCodeLength + 1)
+    {
+        ChipLogError(NotSpecified, "PairDeviceWithCode failed: Invalid pairing payload");
+        return CHIP_ERROR_INVALID_STRING_LENGTH;
+    }
+
+    auto params = Platform::MakeUnique<PairDeviceWithCodeParams>();
+    VerifyOrReturnError(params != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    params->nodeId = nodeId;
+    Platform::CopyString(params->payloadBuffer, sizeof(params->payloadBuffer), payload);
+
+    // Schedule work on the Matter thread
+    return DeviceLayer::PlatformMgr().ScheduleWork(OnPairDeviceWithCode, reinterpret_cast<intptr_t>(params.release()));
```

Comment:
How about using `ScheduleLambda()` instead of passing params via pointer and the reinterpret_casting it again (and the same thing in other places where you are doing `ScheduleWork()`?

Reply:
That’s a great idea! I also noticed this API. Currently, we are using PlatformMgr().ScheduleWork throughout FabricAdmin, but I believe we should standardize on a single scheduling method. I’ve registered the following issue to update everything to use ScheduleLambda() uniformly.

https://github.com/project-chip/connectedhomeip/issues/35900